File: EmarsysMobileShop/EmarsysMobileShop/DataSource.m
//
//  DataSource.m
//  EmarsysMobileShop
//

#import "DataSource.h"

NS_ASSUME_NONNULL_BEGIN

@implementation DataSource

- (nullable instancetype)init {
    NSAssert(false, @"Unavailable init methods was invoked");
    return nil;
}

+ (DataSource *)sharedDataSource {
    static DataSource *_shared = nil;
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
      _shared = [[DataSource alloc] initInternal];
    });
    return _shared;
}

- (instancetype)initInternal {
    self = [super init];
    if (self) {
        _items = [[NSMutableArray alloc] init];
        _categories = [[NSMutableArray alloc] init];
        NSString *path =
            [[NSBundle mainBundle] pathForResource:@"/sample" ofType:@"csv"];
        NSString *content =
            [NSString stringWithContentsOfFile:path
                                      encoding:NSUTF8StringEncoding
                                         error:NULL];
        NSArray<NSString *> *rows = [content
            componentsSeparatedByCharactersInSet:[NSCharacterSet
                                                     newlineCharacterSet]];
        for (int i = 1; i < [rows count]; i++) {
            NSString *row = [rows objectAtIndex:i];
            Item *item = [[Item alloc] init];
            NSArray *cells = [row componentsSeparatedByString:@","];
            for (int j = 0; j < [cells count]; j++) {
                switch (j) {
                case 0:
                    item.itemID = [cells objectAtIndex:j];
                    break;
                case 1:
                    item.link = [cells objectAtIndex:j];
                    break;
                case 2:
                    item.title = [cells objectAtIndex:j];
                    break;
                case 3:
                    item.image = [cells objectAtIndex:j];
                    break;
                case 4:
                    item.category = [cells objectAtIndex:j];
                    if (![_categories containsObject:item.category]) {
                        [_categories addObject:item.category];
                    }
                    break;
                case 5:
                    item.price = [[cells objectAtIndex:j] floatValue];
                    break;
                case 6:
                    item.available = [[cells objectAtIndex:j] boolValue];
                    break;
                case 7:
                    item.brand = [cells objectAtIndex:j];
                    break;
                }
            }
            [_items addObject:item];
        }
    }
    return self;
}

- (NSArray<Item *> *)itemsFromCategory:(NSString *)category {
    NSPredicate *predicate =
        [NSPredicate predicateWithFormat:@"category == %@", category];
    return [_items filteredArrayUsingPredicate:predicate];
}

@end

NS_ASSUME_NONNULL_END

File: EmarsysPredictShopSwift/EmarsysPredictShopSwift/DataSource.swift
//
//  DataSource.swift
//  EmarsysPredictShopSwift
//
//

import Foundation

class DataSource: NSObject {
    
    var categories = [String]()
    var items = [Item]()
    
    private override init() {
        super.init()
    }
    
    class var sharedDataSource: DataSource {
        struct Static {
            static var onceToken: dispatch_once_t = 0
            static var instance: DataSource!
        }
        dispatch_once(&Static.onceToken) {
            let dataSource = DataSource()
            Static.instance = dataSource
            
            let path = NSBundle.mainBundle().pathForResource("/sample", ofType: "csv")!
            let content = try! String(contentsOfFile:path, encoding:NSUTF8StringEncoding)
            let rows = content.componentsSeparatedByCharactersInSet(NSCharacterSet.newlineCharacterSet())
            
            for rowIndex in 1...rows.count-1 {
                let row = rows[rowIndex]
                let item = Item()
                let cells = row.componentsSeparatedByString(",")
                
                for index in 0...cells.count-1 {
                    switch (index) {
                    case 0:
                        item.itemID = cells[index]
                        break
                    case 1:
                        item.link = cells[index]
                        break
                    case 2:
                        item.title = cells[index]
                        break
                    case 3:
                        item.image = cells[index]
                        break
                    case 4:
                        item.category = cells[index]
                        if !dataSource.categories.contains(item.category!) {
                            dataSource.categories.append(item.category!)
                        }
                        break
                    case 5:
                        let s = cells[index] as NSString
                        item.price = s.floatValue
                        break
                    case 6:
                        let s = cells[index] as NSString
                        item.available = s.boolValue
                        break
                    case 7:
                        item.brand = cells[index]
                        break
                    default : break
                    }
                }
                dataSource.items.append(item)
            }
        }
        return Static.instance
    }
    
    func itemsFromCategory(category: String) -> [Item] {
        let filteredArray = items.filter() { $0.category == category }
        return filteredArray
    }
    
}
